packages/flutter_tools/lib/src/fuchsia/fuchsia_device.dart

```diff
@@ -621,8 +659,13 @@ class FuchsiaDevice extends Device {
 
   String _cachedHostAddress;
 
-  /// List the ports currently running a dart observatory.
-  Future<List<int>> servicePorts() async {
+  /// List the ports currently running a dart observatory and the component url
```

Comment:
If possible, I'd suggest pulling the changes to the port discovery logic out into a separate PR, with an issue filed first that describes the problem that it's solving.

OTOH, if it isn't possible to use `session_control` without it, then it would have to stay.

Reply:
If I seperate the changes in port discovery logic to a seperate PR, it seems that any usage of workstation related code will cause flutter_tools to crash. I am not sure if just adding session_tools will pass the tests through. 

Actually without the changes in servicePorts() it is not just `workstation.qemu-x64-release` build will crash `flutter_tools`, even `qemu-x64` build will not be able to connect to dart observatory. I don't understand the root cause of what crashes the original port discovery logic in workstation build, but I think it is because the original port discovery logic does not work correctly when there are multiple instances of flutter_runner running in the vm instance.

It may be possible to land another PR first that contains only `session_control` related changes, and then land this PR on top of that, but then we will get a broken master if we do this. I think test can still pass but real code will break. It won't be a problem if you managed to merge both PRs before anyone fork the HEAD branch through. 

I will create a smaller PR that only consists of session_control changes to see if test will break.

Reply:
session_control should not care about the observatory port.

For port discovery, we are migrating to using inspect to discover the port since it is much more reliable than the hub. There will be a tool built into ffx which does this for you so it might make sense to hold off for now.